storage/clean/oldDatabaseCleaner.go

```diff
@@ -0,0 +1,260 @@
+	if check.IfNil(args.StorageListProvider) {
+		return nil, storage.ErrNilStorageListProvider
+	}
+	if check.IfNil(args.EpochStartNotifier) {
+		return nil, storage.ErrNilEpochStartNotifier
+	}
+	if check.IfNil(args.OldDataCleanerProvider) {
+		return nil, storage.ErrNilOldDataCleanerProvider
+	}
+
+	pathRemoverFunc := func(file string) error {
+		return os.RemoveAll(file)
+	}
+	directoryReader := directoryhandler.NewDirectoryReader()
+
+	odc := &oldDatabaseCleaner{
+		databasePath:           args.DatabasePath,
+		storageListProvider:    args.StorageListProvider,
+		pathRemover:            pathRemoverFunc,
+		directoryReader:        directoryReader,
+		oldDataCleanerProvider: args.OldDataCleanerProvider,
+		oldestEpochsToKeep:     make(map[uint32]uint32),
+	}
+
+	odc.registerHandler(args.EpochStartNotifier)
+
+	return odc, nil
+}
+
+// registerHandler will register a new function to the epoch start notifier
+func (odc *oldDatabaseCleaner) registerHandler(handler EpochStartNotifier) {
+	subscribeHandler := notifier.NewHandlerForEpochStart(
+		odc.epochChangeActionHandler,
+		odc.epochChangePrepareHandler,
+		core.OldDatabaseCleanOrder)
+
+	handler.RegisterHandler(subscribeHandler)
+}
+
+func (odc *oldDatabaseCleaner) epochChangeActionHandler(hdr data.HeaderHandler) {
+	err := odc.handleEpochChangeAction(hdr.GetEpoch())
+	if err != nil {
+		log.Debug("oldDatabaseCleaner: handleEpochChangeAction", "error", err)
+	}
+}
+
+func (odc *oldDatabaseCleaner) epochChangePrepareHandler(_ data.HeaderHandler) {
+}
+
+func (odc *oldDatabaseCleaner) handleEpochChangeAction(epoch uint32) error {
+	newOldestEpoch, err := odc.computeOldestEpochToKeep()
+	if err != nil {
+		return err
+	}
+
+	odc.Lock()
+	odc.oldestEpochsToKeep[epoch] = newOldestEpoch
+	odc.Unlock()
+
+	log.Debug("old database cleaner", "epoch", epoch, "inner map", odc.oldestEpochsToKeep)
+	shouldClean := odc.shouldCleanOldData(epoch, newOldestEpoch)
+	if !shouldClean {
+		return nil
+	}
+
+	err = odc.cleanOldEpochs(epoch)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (odc *oldDatabaseCleaner) shouldCleanOldData(currentEpoch uint32, newOldestEpoch uint32) bool {
+	odc.RLock()
+	defer odc.RUnlock()
+
+	if !odc.oldDataCleanerProvider.ShouldClean() {
+		return false
+	}
+
+	epochForDeletion := currentEpoch - 1
+	previousOldestEpoch, epochExists := odc.oldestEpochsToKeep[epochForDeletion]
+	if !epochExists {
+		log.Debug("cannot delete old databases as the previous epoch does not exist in configuration",
+			"epoch", epochForDeletion,
+			"epochs configuration", odc.oldestEpochsToKeep)
+		return false
+	}
+	if previousOldestEpoch > newOldestEpoch {
+		log.Debug("skipping cleaning of old databases because new oldest epoch is older than previous",
+			"previous older epoch", previousOldestEpoch,
+			"actual older epoch", newOldestEpoch)
+		return false
+	}
+
+	return true
+}
+
+func (odc *oldDatabaseCleaner) computeOldestEpochToKeep() (uint32, error) {
+	odc.Lock()
```

Comment:
you can use odc.RLock() and odc.RUnlock()

Reply:
done